### ISSUE TITLE
A use case with coroutines

### ISSUE THREAD

**Issue by akrzemi1** [akrzemi1/contracts_v3#10]
I think my use case was not recorded. And I am not sure how to phrase it shortly, I would appreciate help.

I want to be able to write a function that returns a C++23 coroutine-based asynchronous generator from the standard library, like `std::generator`, and be able to say that it returns `T`'s that satisfy condition `P`, and at the same time not being required to provide my custom generator.

Here's a longer description of the problem:

Here is a declaration of a function (or coroutine) that returns an awaiter.

```
A coro(X x, Y y);
```

If we called it like this:

```
A a = coro(x1, y1);
```

then `a` is something we can call `co_await a` on. Now, we would like to express a constraint on the values returned by `co_await a`. Type `A`, being an awaiter, has its `operator co_await` in its interface, so we could put the constraint on its return value (i.e., a postcondition). But I think this would compromize the goal for C++23: to have the Standard library providing a handful of reusable Awaiters, so that programmers do not have to write their own. That is, if C++23 Standard Library provides a atd::generator<T> awaiter, then it obviously cannot have a meaningful postcondition in `operator co_await`. Thus, if I want to provide a function/coroutine for returning non-negative `double`s, I cannot just declare:

```
std::generator<double> generate_masses();
```

One option for the programmer would be provide their own Awaiter derived from `std::generator<double>`, and override `operator co_await` only to put a postcondition there.

This would mean you can never use the Library-provided awaitable out of the box if you want to guarantee that the returned values are constrained.

We could think of a solution where a programmer can declare the guarantee on th point of returning the Awaiter:

```
std::generator<double> generate_masses()
  [[await result: result >= 0]];
```

Here `await` is something syntactically similar to a postcondition, but it semantics are: "The funciton/coroutine returns an Awaiter `a`. Each time a value obtained through `co_await a` is produced it satisfies the indicated condition". I do not know if it is implementable, though. This would only work if you `co_await` in the same scope where you create the Awaiter.

**Comment by notadragon** [comment]
I think this isn't specific to coroutines, but is heavily overlapping with the async callback use case more.  You don't want to return generator<double>, you want to return a generator<double> that has additional postconditions on its own functions. 

You could conceivable want to specify the same thing on inputs - say you have a function that takes a generator of positive doubles.   

In truth, I wonder if this should just be added as an additional paragraph in the description for cppapi.contracts.async.

**Comment by sempuki** [comment]
We have api.coroutines, which is the use case, but this example is obviously much more detailed. This is good information and I'd like it some place.

But let's be honest, we're not going to have any solution for coroutines in the first revision, so let's not lose sleep over it.

**Comment by akrzemi1** [comment]
Maybe I failed to capture the core of the issue that I am seeing. Here's what I see. C++20 coroutines are extremely difficult to use for everyday programmers; practically useless due to the amount of code you have to write, and you will probably get it wrong. A normal programmer can only use coroutines if library types are provided that already customize all the 15-or-so customization points and have been thoroughly tested by library vendors. So, the way coroutines have been advertized is that a programmer does not have to worry about the customization points: she just takes a Standard Library (or other library like Boost) component, such as `std::generator<T>` and uses it out of the box. She does not even have to know that there are these customizaiton points, that there is `operator co_await`.

But it turns out that declarations of contracts have not been taken into account when preparing this design. It will not be possible in C++23 to use coroutines "out of the box" and guarantee "postconditions" for values yielded from generators.

**Comment by sempuki** [comment]
I agree with your synopsis. Contracts will need to support coroutines at the function level (my_coro), not the class (std::generator) level. `[[await n > 0]]` is a possible solution. But we're likely going to have zero consensus to do anything about function pointers for async programming. We're going to need to pick our battles. The question is not "how do we fix contracts for coroutines" the question is "will this solution make it impossible to support later?"

**Comment by notadragon** [comment]
I think it's not "not at the class level' it's "also at the class level'.  We have use cases to cover the "contracts for coroutines at the function level' - quite a few of them, as the possibilities for what you might want to specify are much broader than for a simple function.  What we don't have, and what overlaps a lot with callbacks, is the ability to embed contracts into a generic type like std::generator (or, as we do have a use case for, std::function).  I think in theory these are independently solvable, but if both are solved the type-system solution needs to build on the function-specific solution. 

I don't think we'll be able to solve the ease-of-use problems coroutines will be plagued with for the near future with contracts in any reasonable timeframe.
